modules/docs/markdown/06-guides/smithy-transformations.md

```diff
@@ -0,0 +1,124 @@
+---
+sidebar_label: Smithy Transformations
+title: Smithy Transformations
+---
+
+There are times that you may want to transform the Smithy model being used by Smithy4s prior to code generation. In this guide, we will walk through exactly how you can accomplish this. As an example, we will show how you can remove members marked with a certain trait from structures inside of your Smithy model. However, you can use the same principles in this guide to accomplish whatever other transformations you may need.
+
+## Starting Smithy Model
+
+```smithy
+namespace preprocessed
+
+@trait(selector: "structure > member")
+structure removeBeforeCodegen {}
+
+structure MyStruct {
+    @required
+    name: String
+
+    @removeBeforeCodegen
+    id: String
+}
+```
+
+Here we have defined a trait `removeBeforeCodegen`. We have marked the `id` member of `MyStruct` with this trait. As such, we will implement a transformer which will lead to the model looking as follows:
+
+```smithy
+namespace preprocessed
+
+structure MyStruct {
+    @required
+    name: String
+}
+```
+
+This is the model that will ultimately be fed into the Smithy4s code generation tooling.
+
+## Implement the Transformer
+
+Now we will walk through the steps of actually implementing the Smithy model transformer. The `smithy-build` Java library provides an interface for these called `ProjectionTransformer`. Here is an example of a transformer that will remove the members marked with the `removeBeforeCodegen` trait as discussed above.
+
+```scala
+final class RemoveBeforeCodegenTransformation extends ProjectionTransformer {
+  def getName() = {
+    "RemoveBeforeCodegenTransformation"
+  }
+
+  def transform(x: TransformContext) = {
+    val toRemove = x
+      .getModel()
+      .getShapesWithTrait(new ToShapeId() {
+        override def toShapeId(): ShapeId =
+          ShapeId.from("preprocessed#removeBeforeCodegen")
+      })
+    x.getTransformer().removeShapes(x.getModel(), toRemove)
+  }
+}
+```
+
+Here we create a `ProjectionTransformer` named `RemoveBeforeCodegenTransformation`. Inside the `transform` method we remove all shapes that are marked with the `removeBeforeCodegen` trait and return the resultant model.
+
+## Register the Transformer
+
+We need to register the Transformer so that the Smithy tooling will be able to find it when it does a lookup for all `ProjectionTransformer` implementations. We do this by creating a file at `src/main/resources/META-INF/services/` called `software.amazon.smithy.build.ProjectionTransformer`. Inside of this file, we need to place the fully qualified classpath that leads to our transformer (e.g. `preprocessed.transformations.RemoveBeforeCodegenTransformation`).
```

Comment:
Let's add a note saying this is the SPI mechanism from the java stdlib.

Also I'd drop that last sentence, show the actual file contents, and maybe comment on it later with something like

> this file will be used by the [SPI mechanism](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html) used by the Smithy tooling to "register" the class as a transformer that can be used.